File: src/render/decorations.rs
/// Server-side window decorations — title bars with macOS-style controls.
///
/// Each window gets a 30 px title bar rendered above its mapped position.
/// Control dots (left side, macOS-style): close (red), minimize (amber), maximize (green).
use smithay::{
    backend::renderer::{
        element::{solid::SolidColorRenderElement, Id, Kind},
        Color32F,
    },
    desktop::{Space, Window},
    utils::{Physical, Point, Rectangle},
};

pub use crate::shell::layout::DECO_H;

// Title bar background
const BAR_ACTIVE:   Color32F = Color32F::new(0.10, 0.16, 0.38, 0.93);
const BAR_INACTIVE: Color32F = Color32F::new(0.07, 0.10, 0.22, 0.80);

// Control dot colours (close / minimize / maximize)
const BTN_CLOSE: Color32F = Color32F::new(0.878, 0.361, 0.361, 1.0);  // #e05c5c
const BTN_MIN:   Color32F = Color32F::new(0.941, 0.659, 0.188, 1.0);  // #f0a830
const BTN_MAX:   Color32F = Color32F::new(0.361, 0.761, 0.361, 1.0);  // #5cc25c

const BTN_SZ: i32 = 12;   // dot width / height in logical px
const BTN_Y_OFF: i32 = 9; // vertical offset inside 30 px bar: (30 - 12) / 2
const BTN_X0: i32 = 10;   // left margin for close dot
const BTN_GAP: i32 = 6;   // gap between dots

/// Which part of a window decoration was clicked.
#[derive(Clone)]
pub enum DecoHit {
    TitleBar(Window),
    Close(Window),
    Minimize(Window),
    Maximize(Window),
}

/// Return the three button rects (close, min, max) in logical space for a bar at `bar`.
fn btn_rects(bar: Rectangle<i32, smithay::utils::Logical>)
    -> [Rectangle<i32, smithay::utils::Logical>; 3]
{
    let y = bar.loc.y + BTN_Y_OFF;
    [
        Rectangle::new((bar.loc.x + BTN_X0,                       y).into(), (BTN_SZ, BTN_SZ).into()),
        Rectangle::new((bar.loc.x + BTN_X0 + BTN_SZ + BTN_GAP,   y).into(), (BTN_SZ, BTN_SZ).into()),
        Rectangle::new((bar.loc.x + BTN_X0 + (BTN_SZ + BTN_GAP) * 2, y).into(), (BTN_SZ, BTN_SZ).into()),
    ]
}

/// Emit decoration render elements for every window in the space.
///
/// Returns physical-space rectangles (same convention as borders.rs).
/// Push these *before* border and window elements in the render list.
pub fn collect_decorations(
    space: &Space<Window>,
    focused: Option<&Window>,
    output_scale: f64,
) -> Vec<SolidColorRenderElement> {
    let mut out = Vec::new();

    for window in space.elements() {
        let geo = match space.element_geometry(window) { Some(g) => g, None => continue };

        // Bar sits ABOVE the window content
        let bar_log = Rectangle::new(
            (geo.loc.x, geo.loc.y - DECO_H).into(),
            (geo.size.w, DECO_H).into(),
        );

        let is_focused = focused.map(|f| f == window).unwrap_or(false);
        let bar_color  = if is_focused { BAR_ACTIVE } else { BAR_INACTIVE };

        // Helper to convert logical → physical rect
        let to_phys = |r: Rectangle<i32, smithay::utils::Logical>| -> Rectangle<i32, Physical> {
            Rectangle::new(
                ((r.loc.x as f64 * output_scale) as i32, (r.loc.y as f64 * output_scale) as i32).into(),
                ((r.size.w as f64 * output_scale) as i32, (r.size.h as f64 * output_scale) as i32).into(),
            )
        };

        // Title bar background
        out.push(SolidColorRenderElement::new(
            Id::new(), to_phys(bar_log), 0usize, bar_color, Kind::Unspecified,
        ));

        // Control dots
        let [close_r, min_r, max_r] = btn_rects(bar_log);
        let dot_colors = [BTN_CLOSE, BTN_MIN, BTN_MAX];
        for (r, color) in [close_r, min_r, max_r].iter().zip(dot_colors.iter()) {
            out.push(SolidColorRenderElement::new(
                Id::new(), to_phys(*r), 0usize, *color, Kind::Unspecified,
            ));
        }
    }

    out
}

/// Test whether a logical-space pointer position hits a decoration zone.
/// Returns the hit type and the window it belongs to.
pub fn hit_test(space: &Space<Window>, pos: Point<f64, smithay::utils::Logical>) -> Option<DecoHit> {
    let pos_i = Point::<i32, smithay::utils::Logical>::from((pos.x as i32, pos.y as i32));

    for window in space.elements().rev() {
        let geo = match space.element_geometry(window) { Some(g) => g, None => continue };

        let bar = Rectangle::new(
            (geo.loc.x, geo.loc.y - DECO_H).into(),
            (geo.size.w, DECO_H).into(),
        );
        if !bar.contains(pos_i) { continue; }

        let [close_r, min_r, max_r] = btn_rects(bar);
        if close_r.contains(pos_i) { return Some(DecoHit::Close(window.clone())); }
        if min_r.contains(pos_i)   { return Some(DecoHit::Minimize(window.clone())); }
        if max_r.contains(pos_i)   { return Some(DecoHit::Maximize(window.clone())); }

        return Some(DecoHit::TitleBar(window.clone()));
    }
    None
}

File: csharp/Pancake.Render/Decorations.cs
using Pancake.Shell;

namespace Pancake.Render;

/// Which part of a window decoration was clicked.
public abstract record DecoHit
{
    public sealed record TitleBar(PancakeWindow Window) : DecoHit;
    public sealed record Close(PancakeWindow Window) : DecoHit;
    public sealed record Minimize(PancakeWindow Window) : DecoHit;
    public sealed record Maximize(PancakeWindow Window) : DecoHit;
}

// Full port of src/render/decorations.rs -- server-side title bars with
// macOS-style close/minimize/maximize dots. Like borders.rs, this was
// never GPU-side logic (just geometry + a click hit-test), so it was
// blocked purely on PancakeSpace/PancakeWindow, and is now a complete
// port.
public static class Decorations
{
    // Title bar background.
    private static readonly (float R, float G, float B, float A) BarActive = (0.10f, 0.16f, 0.38f, 0.93f);
    private static readonly (float R, float G, float B, float A) BarInactive = (0.07f, 0.10f, 0.22f, 0.80f);

    // Control dot colours (close / minimize / maximize).
    private static readonly (float R, float G, float B, float A) BtnClose = (0.878f, 0.361f, 0.361f, 1.0f); // #e05c5c
    private static readonly (float R, float G, float B, float A) BtnMin = (0.941f, 0.659f, 0.188f, 1.0f);   // #f0a830
    private static readonly (float R, float G, float B, float A) BtnMax = (0.361f, 0.761f, 0.361f, 1.0f);   // #5cc25c

    private const int BtnSz = 12;    // dot width/height in logical px
    private const int BtnYOff = 9;   // vertical offset inside 30px bar: (30-12)/2
    private const int BtnX0 = 10;    // left margin for close dot
    private const int BtnGap = 6;    // gap between dots

    /// Return the three button rects (close, min, max) in logical space for a bar at `bar`.
    private static (Rectangle Close, Rectangle Min, Rectangle Max) BtnRects(Rectangle bar)
    {
        var y = bar.Loc.Y + BtnYOff;
        return (
            new Rectangle(new Point(bar.Loc.X + BtnX0, y), new Size(BtnSz, BtnSz)),
            new Rectangle(new Point(bar.Loc.X + BtnX0 + BtnSz + BtnGap, y), new Size(BtnSz, BtnSz)),
            new Rectangle(new Point(bar.Loc.X + BtnX0 + (BtnSz + BtnGap) * 2, y), new Size(BtnSz, BtnSz))
        );
    }

    /// Emit decoration render elements for every window in the space.
    /// Returns physical-space rectangles (same convention as Borders).
    /// Push these before border and window elements in the render list.
    public static List<BorderElement> CollectDecorations(PancakeSpace space, PancakeWindow? focused, double outputScale)
    {
        var output = new List<BorderElement>();

        Rectangle ToPhys(Rectangle r) => new(
            new Point((int)(r.Loc.X * outputScale), (int)(r.Loc.Y * outputScale)),
            new Size((int)(r.Size.W * outputScale), (int)(r.Size.H * outputScale)));

        foreach (var window in space.Elements())
        {
            var geo = space.ElementGeometry(window);
            if (geo is not { } g) continue;

            // Bar sits ABOVE the window content.
            var barLog = new Rectangle(new Point(g.Loc.X, g.Loc.Y - Layout.DecoH), new Size(g.Size.W, Layout.DecoH));

            var isFocused = focused is not null && focused.Equals(window);
            var barColor = isFocused ? BarActive : BarInactive;

            output.Add(new BorderElement(ToPhys(barLog), barColor.R, barColor.G, barColor.B, barColor.A));

            var (closeR, minR, maxR) = BtnRects(barLog);
            output.Add(new BorderElement(ToPhys(closeR), BtnClose.R, BtnClose.G, BtnClose.B, BtnClose.A));
            output.Add(new BorderElement(ToPhys(minR), BtnMin.R, BtnMin.G, BtnMin.B, BtnMin.A));
            output.Add(new BorderElement(ToPhys(maxR), BtnMax.R, BtnMax.G, BtnMax.B, BtnMax.A));
        }

        return output;
    }

    /// Test whether a logical-space pointer position hits a decoration zone.
    /// Returns the hit type and the window it belongs to.
    public static DecoHit? HitTest(PancakeSpace space, Point pos)
    {
        // Iterate top-to-bottom (reverse stacking order), matching
        // Rust's `space.elements().rev()` -- the topmost window's
        // decoration wins if bars overlap.
        var elements = space.Elements();
        for (var i = elements.Count - 1; i >= 0; i--)
        {
            var window = elements[i];
            var geo = space.ElementGeometry(window);
            if (geo is not { } g) continue;

            var bar = new Rectangle(new Point(g.Loc.X, g.Loc.Y - Layout.DecoH), new Size(g.Size.W, Layout.DecoH));
            if (!bar.Contains(pos)) continue;

            var (closeR, minR, maxR) = BtnRects(bar);
            if (closeR.Contains(pos)) return new DecoHit.Close(window);
            if (minR.Contains(pos)) return new DecoHit.Minimize(window);
            if (maxR.Contains(pos)) return new DecoHit.Maximize(window);

            return new DecoHit.TitleBar(window);
        }
        return null;
    }
}
